File: day-10/part-1/badouralix.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define circle_size 256
#define a *(circle + (current_position + i) % circle_size)
#define b *(circle + (current_position + length - 1 - i) % circle_size)

const int run(char* s)
{
    int circle[circle_size], buffer[circle_size],
        current_position = 0, skip_size = 0, length;

    const char delimiters[] = ",";
    char *token = strtok(s, delimiters);

    for (int i = 0; i != circle_size; i++) {
        circle[i] = i;
    }

    do {
        length = atoi(token);
        for (int i = 0; 2*i + 1 < length; i++) {
            a ^= b ^= a ^= b;
        }
        current_position = (current_position + length + skip_size) % circle_size;
        skip_size++;
    } while (token = strtok(NULL, delimiters));

    return circle[0] * circle[1];
}

int main(int argc, char** argv)
{
    printf("%d\n", run(argv[1]));
    return 0;
}

File: day-10/part-1/badouralix.py
from runners.python import Submission
from copy import copy

class BadouralixSubmission(Submission):

	def run(self, s):
		# :param s: input in string format
		# :return: solution flag
		circle_size = 256
		circle = list(range(circle_size))
		current_position = 0
		skip_size = 0
		lengths = [int(n) for n in s.split(',')]

		for length in lengths:
			circle_copy = copy(circle)
			for i in range(length):
				circle_copy[(current_position + i) % circle_size] = circle[(current_position + length - 1 - i) % circle_size]
			circle = circle_copy
			current_position = (current_position + length + skip_size) % circle_size
			skip_size += 1

		return circle[0] * circle[1]
